### PR TITLE
use dotnet cli to build NETCore

### PR DESCRIPTION
**Pull request by enricosada**
The idea is to build NETCore (coreclr) version using [.NET cli](https://github.com/dotnet/cli).

ref #334

It's complicated to modify the fsproj msbuild targets to build NETCore, the supported way is .NET CLI atm. The dotnetcli can run in parallel to current .fsproj
.NET CLI bundle the F# NETCore (coreclr) compiler, so it's easy to use.

I added some info about f# and NETCore in https://github.com/enricosada/fsharp-dotnet-cli-samples/wiki

After the NETCore assemblies are built, these can be put inside the nuget package created with current rake build script.
So no need to change current build script or fsproj

Roadmap: First building .NET 4.6 (sanity check), adding NETCore target, fixing NETCore, include assemblies in official package.
I think i need some help in `fix NETCore differences`, but that's ok, it's a normal migration from .NET Desktop => .NET Core. You can use compiler directive `DNXCORE50` to ifdef NETCore code

- [X] create .NET CLI build (`project.json`) with framework .NET 4.6
- [X] create test console app referencing Suave `project.json` 
- [x] add appveyor build
- [x] add NETCore (dnxcore50) framework
- [ ] add appveyor build matrix for NETCore (failure allowed)

After that, this pr can be merged, the rest of the work needed is:

- fix NETCore differences (i'll try to help, but i am a bit short of time. Kevin's pr works can help)
- run dotnetcli commands in current rake script
- add NETCore assemblies to nuget package built with rake

## How to build and run a sample test server

After installing .NET CLI, from the root of repository:

- `dotnet restore` restore all packages
- `cd src\Suave.DotnetCLI.Tests` go in the test project
- `dotnet build` build src/Suave and the test console app
- `\bin\Debug\net46\win7-x64\Suave.DotnetCLI.Tests.exe` run the test app

## How to get coreclr assemblies

The assemblies after `dotnet build` are inside subdirectory `bin/{configuration}/{framework}` 
So for example .NET 4.6 are in

```
src\Suave\bin\Debug\net46
```

You can also create a nuget package for suave using .NET CLI, not needed to be used as official, but it's useful to compare with official for deps and binaries

- `cd Suave` 
- `dotnet package` 

```
Producing nuget package "Suave.1.0.0" for Suave
Suave -> d:\github\suave\src\Suave\bin\Debug\Suave.1.0.0.nupkg
Producing nuget package "Suave.1.0.0.symbols" for Suave
Suave -> d:\github\suave\src\Suave\bin\Debug\Suave.1.0.0.symbols.nupkg
```

### CODE REVIEW

**Comment by haf** [comment]
:star2:

**Comment by enricosada** [comment]
ok @haf  i added netcore (dnxcore50).

The compilation fails (it's ok, see below), you can see error in the [appveyor build log](https://ci.appveyor.com/project/AdemarGonzalez/suave/build/1.0.1188#L2429)

to build Suave netcore do:

1. install .NET Cli and clone repo 
2. `cd src\Suave`
3. `dotnet restore`
4. `dotnet build`
5.  the fsc fails with build errors

If you need to change `project.json` redo `restore`, otherwise if you change `.fs` do `build` only.

I fixed all errors i know how to fix, but others are specific of the migration .NET Desktop -> .NET Core of Suave. For example:

- BinaryFormatter
- MemoryCache
- ClaimsPrincipal
- DataContractJsonSerializer

These need some replacement i think in .NET Core or workarounds, dunno. Maybe Kevin pr can help

I dont think i can give a more help there, because some issue are Suave specific ( like BinaryFormatter etc )

How do you want to continue?
 
- i can set appveyor to allow failure on netcore build job, so you can merge this pr and continue to work.
- i can give push access to this pr's branch to someone (you? ).
- pull this in a branch in suave repo

**Comment by enricosada** [comment]
BTW

Remember to revert 363c614 , this commit make appveyor build only NETCore build job, to speedup this pr

Another issue is 2f01fd2 because `Assembly.GetCallingAssembly` can be ok like that or not, depends on your requirerments ( see [this corefx issue](https://github.com/dotnet/corefx/issues/1784#issuecomment-115805990) about Assembly methods )

**Comment by ademar** [comment]
Thanks @enricosada for the head start, much appreciated!
I think we can continue from here; we will be sure to ping you if we need help :)

We will pull this PR into a branch and port Kevin's changes over.

**Comment by enricosada** [comment]
Sure @ademar , feel free to ping me in skype/twitter. Maybe i should add a room in f# slack, we should start to port libraries to netcore (i'll do argu next ) :smile: 

btw i'll add tomorrow this workflow for migration as use case to https://github.com/enricosada/fsharp-dotnet-cli-samples/wiki so feel free to give feedback (wiki is open).

Atm macosx/linux doesnt work so you should do it in windows, i am working on it. 
i'll update https://github.com/enricosada/fsharp-dotnet-cli-samples when ready with a working  .travis.yml too

About Kevin pr, you should port only f# code changes (like Close -> Dispose, etc), the project.json in this pr is ok

Pls ping me when you have a `dotnet build` green

**Comment by ademar** [comment]
Boys everything is green here cc @enricosada @haf

**Comment by haf** [comment]
Read through and looks good to me, besides the point about assembly version above.

File: src/Suave/project.json [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        "SuaveConfig.fs",
+        "ParsingAndControl.fs",
+        "Web.fs",
+        "Owin.fs",
+        "WebSocket.fs",
+        "Json.fs"
+    ],
+
+    "frameworks": {
+        "dnxcore50": {
```
Comment (davidfowl) [inline src/Suave/project.json]: Libraries shouldn't target this framework. Use dotnet5.x instead (which will be netstandard very soon)

**Pull request closed without merging**